### PR TITLE
revise metrics and cumulative docstrings/APIs

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

### Description
this is mostly non-breaking, updated the base cumulative API to have `extend` and `append`. fixed multiple docstrings to be clear about the shape assumptions.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/metrics/metric.py [hunk 1]
```diff
@@ -15,121 +15,165 @@
...
+
+    The following is an example of extending with variable length data:
+
+    .. code-block:: python
+
+        import torch
+        from monai.metrics import Cumulative
+
+        c = Cumulative()
+        c.extend(torch.zeros((8, 2)), torch.zeros((6, 2)))  # adds batches
```
Comment (Nic-Ma) [inline monai/metrics/metric.py:160]: Could you please help explain more about the usage of `extend`?
I didn't get why adding this new API against the `append` or the previous `add`.

Thanks.

File: monai/metrics/metric.py [hunk 2]
```diff
@@ -15,121 +15,165 @@
...
     def _compute_list(self, y_pred: TensorOrList, y: Optional[TensorOrList] = None):
         """
-        Excute the computation for the y_pred and y items of a iteration, the data is in the list shape.
-        Will concat the results to guarantee the output shape of ret is BCHW[D], otherwise it's list of batch-first,
-        which is against our principle that data in metrics should be BCHW[D] or list of channel-first.
-        Note: subclass may enhance the operation with multi-threads to accelerate.
+        Execute the metric computation for `y_pred` and `y` in a list of "channel-first" tensors.
 
+        The return value is a "batch-first" tensor, or a list of "batch-first" tensors.
+        When it's a list of tensors, each item in the list can represent (batched) class-wise metrics.
```
Comment (Nic-Ma) [inline monai/metrics/metric.py:82]: I think usually the multi-class metrics is a multi-channel Tensor, not items of the list, here if every item of `ret` is a list, means the metric returns multiple items for every iteration, for example:
```py
class MyMetric:
    ...
    def _compute_tensor(self):
        # return some median computation result for every iteration, will aggregate for the final value later
        return TP, FP, WD
```

What do you think?

Thanks.

File: monai/metrics/metric.py (hunk 1 above)
Comment (wyli) [reply monai/metrics/metric.py:160]: yes, these APIs will provide flexibility in handling the batch dimension.

File: monai/metrics/metric.py [hunk 3]
```diff
@@ -140,64 +184,103 @@ def reset(self):
...
+                they will be concatenated at the 0-th dimension when `get_buffer()` is called.
+        """
+        if self._buffers is None:
+            self._buffers = [[] for _ in data]
+        for b, d in zip(self._buffers, data):
+            # converting to pytorch tensors so that we can use the distributed API
+            d_t: torch.Tensor
+            d_t, *_ = convert_data_type(d, output_type=torch.Tensor, wrap_sequence=True)  # type: ignore
+            try:
+                b.extend([x[0] for x in torch.split(d_t, 1, dim=0)])
```
Comment (Nic-Ma) [inline monai/metrics/metric.py:203]: I doult there may be a little performance drop to split the batch into list and stack them in `sync()`, the previous code store `batch-first` Tensor into buffer directly and concat several `batch-first` Tensors in `sync()`.
Maybe you can do some quick performance test if batch size is bigger?

Thanks in advance.

Comment (wyli) [reply monai/metrics/metric.py:203]: Hi Nic, this is to have consistent buffer elements in the implementation, and the base class is to show the concepts. If the performance is an issue, the subclass can provide other implementations.

File: monai/metrics/metric.py [hunk 4]
```diff
@@ -15,121 +15,168 @@
...
-            y_pred: the model prediction data to compute, must be a list of `channel-first` Tensor
+            y_pred: the raw model prediction data at one iteration, must be a list of `channel-first` Tensor
                 or a `batch-first` Tensor.
             y: the ground truth to compute, must be a list of `channel-first` Tensor
                 or a `batch-first` Tensor.
 
+        Returns:
+            The computed metric values at the iteration level.
+            The output shape could be a `batch-first` tensor or a list of `batch-first` tensors.
+            When it's a list of tensors, each item in the list can represent (batched) class-wise metrics.
```
Comment (Nic-Ma) [inline monai/metrics/metric.py:62]: Could you please help also update this line to `a specific type of metric values.`?

Thanks.

File: monai/metrics/metric.py [hunk 5]
```diff
@@ -140,64 +187,106 @@ def reset(self):
...
+                they will be concatenated at the 0-th dimension when `get_buffer()` is called.
+        """
+        if self._buffers is None:
+            self._buffers = [[] for _ in data]
+        for b, d in zip(self._buffers, data):
+            # converting to pytorch tensors so that we can use the distributed API
+            d_t: torch.Tensor
+            d_t, *_ = convert_data_type(d, output_type=torch.Tensor, wrap_sequence=True)  # type: ignore
+            try:
+                b.extend([x[0] for x in torch.split(d_t, 1, dim=0)])
```
Comment (Nic-Ma) [inline monai/metrics/metric.py:207]: I think maybe `torch.unbind()` is more suitable than `torch.split` here?

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update.
Looks good to me, just some minor comments.

Thanks.

File: monai/metrics/metric.py [hunk 6]
```diff
@@ -79,9 +77,14 @@ def _compute_list(self, y_pred: TensorOrList, y: Optional[TensorOrList] = None):
         Execute the metric computation for `y_pred` and `y` in a list of "channel-first" tensors.
 
         The return value is a "batch-first" tensor, or a list of "batch-first" tensors.
-        When it's a list of tensors, each item in the list can represent (batched) class-wise metrics.
+        When it's a list of tensors, each item in the list can represent a specific type of metric values.
 
-        Note: subclass may enhance the operation with multi-threads to accelerate.
+        For example, `self._compute_tensor` may be implemented as returning a list of `batch_size` items,
```
[the remark this answers is not in the document]
Comment (wyli) [reply monai/metrics/metric.py:82]: Hi Nic, this is to show the number of items in the list is `batch_size`, and in the later docstring each item has the length of `batch_size`.

[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/metrics/metric.py:82]: OK, I see.
Maybe you can try to make it more clear to avoid confusion, some thing like:
`returning a list of `batch-first` items, length of each item is same as `batch_size`?

Thanks in advance.

File: monai/metrics/metric.py (hunk 5 above)
Comment (wyli) [reply monai/metrics/metric.py:207]: Not sure why we make the change, could you please clarify the advantage?

File: monai/metrics/metric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wyli) [reply monai/metrics/metric.py:82]: I think it's clear that I use "`" to denote the size, and the double-quote for the word"batch_first"

File: monai/metrics/metric.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/metrics/metric.py:207]: I think `torch.split()` is mainly used for specified length of data chunks, here we fallback to use `length=1`, seems `torch.unbind` is simpler for this use case?

Thanks.

File: monai/metrics/metric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/metrics/metric.py:82]: OK, I missed line 79.

Thanks.

File: monai/metrics/metric.py (hunk 5 above)
Comment (wyli) [reply monai/metrics/metric.py:207]: It will change the outcome for one of the test case:
`[x[0] for x in torch.split(torch.tensor([]), 1, dim=0)]` vs `torch.unbind(torch.tensor([]), 0)`, I think the split is better here.

Comment (Nic-Ma) [reply monai/metrics/metric.py:207]: OK.

Thanks.
